### PR TITLE
Adds a new Widget Inspector API: `getSelectedLocalWidget`

### PR DESCRIPTION
**Pull request by elliette**
Work towards https://github.com/flutter/devtools/issues/8626

When DevTools is notified that a selection change has occurred (i.e., a user has selected a widget using the on-device widget inspector), it then sends a request to Flutter to get the selected widget (`getSelectedWidget`). 

This PR adds a new API that DevTools can call (`getSelectedLocalWidget`) when implementation widgets are hidden in the tree. `getSelectedLocalWidget` will return the currently selected widget if it was created in the user's project, or will return the nearest ancestor of the currently selected widget if its an implementation widget.

After this has landed, we can follow up with the necessary DevTools change (Draft PR here: https://github.com/flutter/devtools/pull/8625)

### CODE REVIEW

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 1]
```diff
@@ -1293,12 +1293,16 @@ mixin WidgetInspectorService {
       callback: _getSelectedWidget,
       registerExtension: registerExtension,
     );
+    _registerServiceExtensionWithArg(
+      name: WidgetInspectorServiceExtensions.getSelectedLocalWidget.name,
+      callback: _getSelectedLocalWidget,
+      registerExtension: registerExtension,
+    );
     _registerServiceExtensionWithArg(
       name: WidgetInspectorServiceExtensions.getSelectedSummaryWidget.name,
```
Comment (kenzieschmoll) [inline packages/flutter/lib/src/widgets/widget_inspector.dart:1302]: how does this new api differ from this one?

**Comment by kenzieschmoll** [comment]
Another question: can we walk up the tree on the DevTools client instead? Since "show implementation widgets" is a DevTools setting, it seems like we should handle the tree walking on our side instead of in the framework.

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 1 above)
Comment (elliette) [reply packages/flutter/lib/src/widgets/widget_inspector.dart:1302]: Hahaha I was very tunnel visioned and did not notice this. Looking at the code, I'm guessing they do the same thing. Let me give that one a try and if so I will close this PR 🙃

**Comment by elliette** [comment]
> Another question: can we walk up the tree on the DevTools client instead? Since "show implementation widgets" is a DevTools setting, it seems like we should handle the tree walking on our side instead of in the framework.

No we can't unfortunately, because we build the parent-child relationships when we build the tree itself. When we get a selected implementation widget but implementation widgets are hidden, we have no way of getting its parent and walking up the tree.

File: packages/flutter/lib/src/widgets/widget_inspector.dart (hunk 1 above)
Comment (elliette) [reply packages/flutter/lib/src/widgets/widget_inspector.dart:1302]: `getSelectedSummaryWidget` works exactly the same as this new API, so I will close this PR 🤦‍♀️

**Pull request closed without merging**